### ISSUE TITLE
LZ4_decompress_safe() failure and wrong return value?

### ISSUE THREAD

**Issue by soumyadeep2007** [lz4/lz4#859]
**Describe the bug**
`LZ4_decompress_safe()` fails when given the true size of `dst` as `dstCapacity`. However, inspired by #421, if we pass in `dstCapacity + 1` in the function call, it succeeds and returns `dstCapacity` (note that we did not even need to allocate that one extra byte to `dst`). PFA lz4_decompress_issue.zip, which has a small program to reproduce and a binary dump of `src` taken with gdb.

**To Reproduce**
1. Place files attached in /tmp

2. Compile and run demo.c
```
clang /tmp/demo.c -llz4
/tmp/a.out
```

demo.c:

```c
#include <stdio.h>
#include <unistd.h>
#include <lz4.h>

int main()
{
	FILE *f;
	char src[8112] = {0};
	char dst[631811] = {0};

	f = fopen("/tmp/decompress_issue.bin", "rb");

	if(!f)
		return 1;
	else
		fread(src, 8112, 1, f);

	printf("%d\n", LZ4_decompress_safe(src, dst, 8112, 631811)); // -8108
	printf("%d\n", LZ4_decompress_safe(src, dst, 8112, 631811 + 400)); // 631811
	printf("%d\n", LZ4_decompress_safe(src, dst, 8112, 631811 + 4)); // 631811
	printf("%d\n", LZ4_decompress_safe(src, dst, 8112, 631811 + 1)); // 631811

	return 0;
}

```

**Expected behavior**
`LZ4_decompress_safe()` should pass with `dstCapacity = 631811`, which is the true size of the uncompressed data. If `src` is malformed, it should not pass with `dstCapacity + <x>`

**Desktop (please complete the following information):**
 - OS: 
```
cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=19.04
DISTRIB_CODENAME=disco
```
 - Version : 1.9.2
 - Compiler:
```
clang --version
clang version 8.0.0-3 (tags/RELEASE_800/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
```
 - Build System - make
 - Other hardware specs :  Intel(R) Xeon(R) CPU E5-2630 v3 @ 2.40GHz, 32 CPUs, Little Endian, x86_64

[lz4_decompress_issue.zip](https://github.com/lz4/lz4/files/4595376/lz4_decompress_issue.zip)

**Comment by soumyadeep2007** [comment]
This occurs not only in the released 1.9.2 version (as reported above), but we have encountered this in the `dev` branch as well (f39b79fb02962a1cd880bbdecb6dffba4f754a1) and also in 1.9.1 and also 1.7.1

**Comment by BellaXlp** [comment]
@soumyadeep2007

**Comment by BellaXlp** [comment]
@soumyadeep2007 
There is nothing wrong with the lz4 code.
The problem is that your compressed file(decompress_issue.bin) is not correct. There is one byte missing from the compressed file.

The attachment code can be verified.
[demo_modify.zip](https://github.com/lz4/lz4/files/4598782/demo_modify.zip)

**Comment by soumyadeep2007** [comment]
@BellaXlp thanks for looking into this for us.

@d and I looked into this some more.

First, we wanted to ensure that our expectations of the API are correct.

Say, we want to `LZ4_compress_destSize` all of 100 bytes of X which is of 100 bytes into a 50 byte destination and the result is Y.

and say we get a result saying that 90 of the 100  bytes have been compressed successfully and 45 bytes of the destination has been used

q1) can we assume that we can `LZ4_decompress_safe` Y of size 45 bytes into a 90 byte buffer?
(In other words, can the `srcSizePtr` returned as an output parameter from `LZ4_compress_destSize` be used for `dstCapacity` in `LZ4_decompress_safe`?)


q2) can we assume that we can LZ4_decompress_safe Y of size 45 bytes into a 100 byte buffer?
(In other words, can the input value of `srcSizePtr` passed into `LZ4_compress_destSize`  be used for `dstCapacity` in `LZ4_decompress_safe`?)

Attached is a demo (**done with 1.8.3**), that contains a golden copy of the uncompressed data, a golden copy of the compressed data. We demonstrate how a round-trip of a compress-decompress yields surprising results. Giving the decompression step 1 extra byte for dstCapacity causes it to pass and yield an identical copy of the  golden copy of the uncompressed data.

Our demo seems to suggest that the answer to  q1) is no. But we want to hear it from you!

[demo1.c.txt](https://github.com/lz4/lz4/files/4624253/demo1.c.txt)

**Comment by BellaXlp** [comment]
Q1 and q2 are both ok.

**Comment by soumyadeep2007** [comment]
Hey, can you please help me and @d out with our questions above when you get a chance?

Thanks!

**Comment by Cyan4973** [comment]
The answer to both questions is yes : 
`LZ4_decompress_safe()` is supposed to work with a destination buffer which is _exactly_ the decompressed size, and a destination buffer which is _larger_ than decompressed size.

Now, @BellaXlp seems to imply that the expected decompressed size is wrong. I haven't looked myself at this question, but I'll assume he is right, as it would explain why `LZ4_decompress_safe()` needs a larger buffer than the one your are expecting.

Then, the second presentation of the failure scenario implies that the source size is actually provided by `LZ4_compress_destSize()`, which tells how many bytes it could compress in a given budget.

So one possibility is that maybe there is a bug in `LZ4_compress_destSize()`, returning a wrong off-by-1 value within `*srcSizePtr`.

In order to be more conclusive, I would have to look deeper into a reproduction case.

Unfortunately, my current availability is so reduced, it's difficult to commit to any date.

**Comment by soumyadeep2007** [comment]
@Cyan4973 thanks for your reply! 


> Then, the second presentation of the failure scenario implies that the source size is actually provided by `LZ4_compress_destSize()`, which tells how many bytes it could compress in a given budget.

> In order to be more conclusive, I would have to look deeper into a reproduction case.

Thanks! Attached is [demo2.c.txt](https://github.com/lz4/lz4/files/4720354/demo2.c.txt), which is a bit more minimal version of demo1.c.txt. Hopefully, it will help.

So essentially we try to compress a byte array (embedded in the file) of size 614618 into a buffer of size 8112 using `LZ4_compress_destSize()`. 614509 bytes were compressed successfully and all of the 8112 bytes were filled up. 
- When we now tell `LZ4_decompress_safe` that the uncompressed size of this 8112 compressed byte array is **614509**, it **fails**. (q1)
- When we now tell `LZ4_decompress_safe` that the uncompressed size of this 8112 compressed byte array is **614618**, it **fails**. (q2)
- When we now tell `LZ4_decompress_safe` that the uncompressed size of this 8112 compressed byte array is **614619**, it **succeeds** and the result is identical to the uncompressed data we started with. (q2)

Output of running demo2.c.txt:
```
compress: 8112
resultant srcSize from compress step: 614509
decompress with dstCapacity = resultant srcSize (614509) from compress step: -8108
decompress with dstCapacity = uncompressed_len (614618): -8108
decompress with dstCapacity = uncompressed_len + 1 (614619): 614618
memcmp decompression output against golden uncompressed data: 0
```

> Unfortunately, my current availability is so reduced, it's difficult to commit to any date.

Totally understand!

**Comment by Cyan4973** [comment]
Thanks for the reproduction case.

The way I understand it, there is probably a bug in `LZ4_compress_destSize()`.
I will have to spend the time to properly trace the scenario and get to the root of the problem.

**Comment by Cyan4973** [comment]
OK, so I've been spending quite some time on this `demo2.c.txt` reproduction case from @soumyadeep2007 

Initially, I misinterpreted what the test case was doing : 
since it invokes `LZ4_compress_destSize()` followed by `LZ4_decompress_safe()`,
I thought it was doing a round-trip compression-decompression of some golden content, 
just providing different buffer size parameters to illustrate the different outcome.

But I missed that the reproduction case actually shipped its own golden compressed content,
and used that instead of the outcome of `LZ4_compress_destSize()` during decompression.
This lead to confusing conclusions.

After the test got updated to perform a round-trip, I got some pretty different results.
Using the golden content of 614618 bytes, to be compressed into a max budget of 8112 bytes, I measure : 

v1.7.3, v1.7.4, v1.7.5, v1.8.0, v1.8.1 : 
- 614509 bytes are compressed into 8112 bytes
- when decompressing into a buffer of 614509 bytes : success, decompress 614509 bytes
- when decompressing into a buffer of 614618 bytes : success, decompress 614509 bytes
- when decompressing into a buffer of 614619 bytes : success, decompress 614509 bytes

v1.8.2, v1.8.3, v1.9.0, v1.9.1 : 
- 614617 bytes are compressed into 8112 bytes
- when decompressing into a buffer of 614617 bytes : __failure__
- when decompressing into a buffer of 614618 bytes : success, decompress 614617 bytes
- when decompressing into a buffer of 614619 bytes : success, decompress 614617 bytes

v1.9.2, dev : 
- 614509 bytes are compressed into 8112 bytes
- when decompressing into a buffer of 614509 bytes : success, decompress 614509 bytes
- when decompressing into a buffer of 614618 bytes : success, decompress 614509 bytes
- when decompressing into a buffer of 614619 bytes : success, decompress 614509 bytes

From `v1.8.2` to `v1.9.1`, something different was happening, nominally compressing more in the same budget, but triggering some issue during decompression. My guess is that these versions produce actually a wrong outcome by one byte: they likely only pack 4 literals at the end, instead of the mandatory minimum of 5 literals they are supposed to.

That being said, it seems that we are now back to normal : v1.9.2 and `dev` work properly, as they used to before `v1.8.2`.

It looks to me that there is no problem with `lz4` `v1.9.2` and `dev`, at least with regards to this golden content.
The reported errors are related to the fact that the golden compressed content itself is incorrect (as detected by @BellaXlp).

Btw, I'm unsure how this golden content was generated. 
None of the tested versions of `lz4` ever produced a compressed block equal to the golden compressed content.

If anyone is interested, I updated the reproduction case, and published it at : 
https://gist.github.com/Cyan4973/358a6853ca4a79ccbc989029c486e658 .
This updated version correctly tests the round trip, 
and alongside compares the produced compressed result to the golden compressed content.
I ran it with multiple versions of `liblz4`.
All other variables from the test stay the same.

@soumyadeep2007 :
in order to continue this investigation, I would need : 
- An explanation on how the golden compressed content was created
- If the issue is considered present in `v1.9.2` : a different golden content, able to trigger a problem in the round-trip test between `LZ4_compress_destSize()` and `LZ4_decompress_safe()`, if such a content exists.

Note :
I will add some comments in the code to document the `LZ4_compress_destSize()` bug between `v1.8.2` and `v1.9.1`. 
It can still be an issue in situations and systems where the version of the compressor cannot be ensured.

**Comment by Cyan4973** [comment]
Closing.
The issue seems to be a wrongly formatted block, resulting in a detected corruption error at decompression time.
The generation issue seems fixed since v1.9.2.

**Comment by Lveyou** [comment]
I think "If the source stream is detected malformed" should return a minimum int:

```
#define LZ4_MALFORMED_SOURCE INT_MIN

int ret = LZ4_decompress_safe(...) ;
if(ret < 0 )
{
    if(ret == LZ4_MALFORMED_SOURCE)
        ;
    else
        ;
}
```
